resolwe/flow/managers/state.py

```diff
@@ -0,0 +1,234 @@
+"""Synchronized singleton-state container and runtime constants for the manager's needs."""
+
+# This module should not import anything local, or there will be circular dependencies,
+# since the constants are needed in various sub-modules inside resolwe.flow.managers.
+
+import json
+from collections import namedtuple
+
+import redis
+
+from django.conf import settings
+
+ManagerChannelPair = namedtuple('ManagerChannelPair', ['queue', 'queue_response'])
+
+MANAGER_CONTROL_CHANNEL = 'control'
+MANAGER_EXECUTOR_CHANNELS = ManagerChannelPair('queue', 'queue_response')
+MANAGER_STATE_PREFIX = 'state_prefix'
+MANAGER_SYNC_PIPE = '_sync_pipe'
+
+
+def update_constants():
+    """Recreate channel name constants to take into account settings changes."""
+    global MANAGER_CONTROL_CHANNEL, MANAGER_EXECUTOR_CHANNELS  # pylint: disable=global-statement
+    global MANAGER_STATE_PREFIX, MANAGER_SYNC_PIPE  # pylint: disable=global-statement
+    redis_prefix = settings.FLOW_MANAGER['REDIS_PREFIX']
+
+    MANAGER_CONTROL_CHANNEL = redis_prefix + '.control'
+    MANAGER_EXECUTOR_CHANNELS = ManagerChannelPair(
+        redis_prefix + '.result_queue',
+        redis_prefix + '.result_queue_response',
+    )
+    MANAGER_STATE_PREFIX = redis_prefix + '.state'
+    MANAGER_SYNC_PIPE = redis_prefix + '._sync_pipe'
+
+
+update_constants()
+
+
+def _with_pseudosetter(prop):
+    """Wrap a property declared as read-only to include a setter which only passes through the same object."""
+    prop_name = '_' + prop.fget.__name__
+
+    def _pseudosetter(self, val):
+        """Set new property value; pseudo-setter, since we don't actually want to allow new values."""
+        if getattr(self, prop_name) is not val:
+            raise AttributeError("can't set attribute")
+    return prop.setter(_pseudosetter)
+
+
+class ManagerState(object):
+    """State interface implementation.
+
+    This takes care of operation atomiticy and synchronization between manager workers.
```

Comment:
Ok, so this class basically supports executing a set of expressions (maths, compare-and-swap etc.) on the state backed by Redis. And using these expressions then requires "setter forwarding" (e.g., pseudosetter decorator).

Can you document this a little better here? Also I am not sure if I like mixing the actual Redis expression interface (e.g., LuaFunction, RedisAtomicBase, IntegerDatum, ObjectDatum), which can be used in a general way, with the actual manager state (e.g., executor_count, sync_semaphore, etc.). I believe splitting these two would make the code easier to follow.

Reply:
The nested classes are sort of generic, but only in the sense of providing "generalized" support for the mechanics of the variables instantiated from them. In the end, this "general" support is still extremely specific to what `ManagerState` needs to do. Until, if ever, there is need for such functionality elsewhere, I think it would be unhealthy to factor them out, since they currently have no other uses whatsoever. Separating them would imply a sense of usefulness which simply isn't there.

I've extended the docstrings to describe the situation better, especially `ManagerState`'s and for the pseudo-setter wrapper.